PaC 101: Getting Started with PaC (#2421)
pulumi/docs: content/blog/getting-started-with-pac/index.md

---

> title: "Getting Started With Pac"

Reviewer: Should this be `PaC`? Do we want to spell it out like `Policy as Code`? Or `Policy as Code (PaC)`?

Author: fixed

date: 2020-02-11
meta_desc: "Pulumi CrossGuard validates infrastructure before deployment to prevent data breaches caused by misconfigured services, "

> meta_image: crossguard.svg

Reviewer: Does Twitter support displaying .svg?

Reviewer: Also, if we're not using the `meta.png` file, it should be deleted from this directory.

Author: It has on previous tweets, meta.png deleted.

authors:
    - sophia-parafina
tags:
    - "Policy as Code"
    - S3
    - Elasticsearch
    - "data breach"
---

Modern applications have brought many benefits and improvements, including the ability to scale and rapid iterations to update software. However, this has come at the cost of complexity. Modern infrastructure is composed of many resources that require detailed configuration to work correctly and securely. Even managed solutions from cloud service providers need additional configuration to ensure that services are secure and free of defects. Cloud providers, such as AWS, do allow you to create policies to ensure that applications are secure, but they are specific to resources that are already deployed. A significant benefit of Policy as Code is the ability to verify and spot problems before deploying your infrastructure.

<!--more-->

> ## Leaky  Buckets

Reviewer 2: ```suggestion
## Leaky Buckets
```

What do many of the most innovative companies in the modern cloud era have in common? They’ve all had massive data breaches from leaky AWS S3 Buckets.  One of the most common reasons for data breaches is misconfigured AWS S3 Buckets. Both security professionals and bad actors use tools such as slurp to find open S3 buckets by just entering a domain, keyword, or even AWS credentials with the slurp CLI. Ideally, we want to proactively prevent public access to our S3 buckets before they are created.

![S3 buckets open to public](Slurp-Amazon-AWS-S3-Bucket-Enumerator.png)

> Pulumi's CrossGuard helps customers adhere to security best practices by checking the final state of a deployment against a set of policies and revealing any non-compliant services. CrossGuard is an SDK for writing policies for services, and it includes best practices playbooks for cloud providers such as AWS, Azure, and Google Cloud Platform. The [Getting Started with PaC]({{< ref "/docs/get-started/crossguard" >}}) covers the case of leaking S3 buckets, but let’s do a quick review with an example showing how we can detect a leaky bucket before it gets deployed.

Reviewer: ```suggestion
Pulumi's CrossGuard helps customers adhere to security best practices by checking the final state of a deployment against a set of policies and revealing any non-compliant services. CrossGuard is an SDK for writing policies for services, and it includes best practices playbooks for cloud providers such as AWS, Azure, and Google Cloud Platform. The [Getting Started with PaC]({{< relref "/docs/get-started/crossguard" >}}) covers the case of leaking S3 buckets, but let’s do a quick review with an example showing how we can detect a leaky bucket before it gets deployed.
```

> >Note: If you haven’t installed Pulumi, follow the [Pulumi Getting Started]({{< ref "/docs/get-started/install" >}})

Reviewer: ```suggestion
> Note: If you haven’t installed Pulumi, follow the [Pulumi Getting Started]({{< relref "/docs/get-started/install" >}})
```

Start by creating an S3 bucket with Pulumi.

```bash
$ pulumi new aws-typescript
```

 Running `pulumi new` creates the project files as well as the code to create a single S3 bucket. We’ll use an S3 ACL to make the bucket public by setting it to the [canned ACL `public-read`](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl).

```ts
const bucket = new aws.s3.Bucket("my-bucket", {
    acl: "public-read",
});
```

Next, we’ll create a Policy Pack, by creating a directory and making it the current working directory.

```bash
$ mkdir policypack && cd policypack
```

CrossGuard is still in *preview*, and we need to set the environmental variable `PULUMI_EXPERIMENTAL` to use it. We can prepend it before creating the policy.

```bash
$ PULUMI_EXPERIMENTAL=true pulumi policy new aws-typescript
```

By default, this creates a policy that checks to see if the ACL sets the bucket to be publically accessible.

```ts
new PolicyPack("aws-typescript", {
    policies: [{
        name: "s3-no-public-read",
        description: "Prohibits setting the publicRead or publicReadWrite permission on AWS S3 buckets.",
        enforcementLevel: "mandatory",

>         validateResource: validateTypedResource(aws.s3.Bucket, (bucket, args, reportViolation) => {

Reviewer: Note: We recently renamed `validateTypedResource` to `validateResourceOfType` (though, `validateTypedResource` can still be used -- it's just marked as deprecated).

If this is published after https://github.com/pulumi/templates-policy/pull/15 is merged, we should change it here:

```suggestion
        validateResource: validateResourceOfType(aws.s3.Bucket, (bucket, args, reportViolation) => {
```

Otherwise, I'll come back and take care of updating the blog post after https://github.com/pulumi/templates-policy/pull/15 is merged.

            if (bucket.acl === "public-read" || bucket.acl === "public-read-write") {
                reportViolation(
                    "You cannot set public-read or public-read-write on an S3 bucket. " +
                    "Read more about ACLs here: https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html");
            }
        }),
    }],
});
```

> A policy consists of a name; a description that is displayed when the policy is triggered; an enforcement level of either `advisory` or `mandatory`; and `validateResource` which specifies the type of resource, the resource, any arguments, and the message to report the violation.

Reviewer: ```suggestion
A policy consists of a name; a description that is displayed when the policy is violated; an enforcement level of either `advisory` or `mandatory`; and `validateResource` which specifies the type of resource, the resource, any arguments, and the message to report the violation.
```

We’ll take the same approach that we used to secure S3 buckets. First, we'll create an Elasticsearch deployment, write a policy, then apply it.

```bash
$ pulumi new aws-typescript
```

Replace `index.ts` with the following to deploy Elasticsearch:

```ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const example = new aws.elasticsearch.Domain("example", {
   clusterConfig: {
       instanceType: "r4.large.elasticsearch",
   },
   elasticsearchVersion: "1.5",
   snapshotOptions: {
       automatedSnapshotStartHour: 23,
   },
   tags: {
       Domain: "TestDomain",
   },
});
```

> In this example, we’ll use [AWSGuard]({{< ref "/docs/guides/crossguard/awsguard.md" >}}), which is a set of best practice policies for AWS. First, create a directory, change into it, and create a new policy.

Reviewer: ```suggestion
In this example, we’ll use [AWSGuard]({{< relref "/docs/guides/crossguard/awsguard" >}}), which is a set of best practice policies for AWS. First, create a directory, change into it, and create a new policy.
```

```bash
$ mkdir awsguard && cd awsguard
$ PULUMI_EXPERIMENTAL=true pulumi policy new awsguard-typescript
```

> The default policy enforcement level is`advisory,` but given the frequency of misconfigured Elasticsearch deployments, we’ll set to `mandatory`.

Reviewer: ```suggestion
The default policy enforcement level is `advisory,` but given the frequency of misconfigured Elasticsearch deployments, we’ll set to `mandatory`.
```
